# Prosper Loan Data Exploration

In [25]:
#Import normal packages and set visulizations to be inline
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from pandas.api.types import CategoricalDtype
%matplotlib inline

## Load Dataset

In [2]:
df = pd.read_csv('prosperLoanData.csv')
df.head()

ListingKey  ListingNumber            ListingCreationDate  \
0  1021339766868145413AB3B         193129  2007-08-26 19:09:29.263000000   
1  10273602499503308B223C1        1209647  2014-02-27 08:28:07.900000000   
2  0EE9337825851032864889A          81716  2007-01-05 15:00:47.090000000   
3  0EF5356002482715299901A         658116  2012-10-22 11:02:35.010000000   
4  0F023589499656230C5E3E2         909464  2013-09-14 18:38:39.097000000   

  CreditGrade  Term LoanStatus           ClosedDate  BorrowerAPR  \
0           C    36  Completed  2009-08-14 00:00:00      0.16516   
1         NaN    36    Current                  NaN      0.12016   
2          HR    36  Completed  2009-12-17 00:00:00      0.28269   
3         NaN    36    Current                  NaN      0.12528   
4         NaN    36    Current                  NaN      0.24614   

   BorrowerRate  LenderYield    ...     LP_ServiceFees  LP_CollectionFees  \
0        0.1580       0.1380    ...            -133.18                0.0   
1        0.0920       0.0820    ...               0.00                0.0   
2        0.2750       0.2400    ...             -24.20                0.0   
3        0.0974       0.0874    ...            -108.01                0.0   
4        0.2085       0.1985    ...             -60.27                0.0   

   LP_GrossPrincipalLoss  LP_NetPrincipalLoss LP_NonPrincipalRecoverypayments  \
0                    0.0                  0.0                             0.0   
1                    0.0                  0.0                             0.0   
2                    0.0                  0.0                             0.0   
3                    0.0                  0.0                             0.0   
4                    0.0                  0.0                             0.0   

   PercentFunded  Recommendations InvestmentFromFriendsCount  \
0            1.0                0                          0   
1            1.0                0                          0   
2            1.0                0                          0   
3            1.0                0                          0   
4            1.0                0                          0   

  InvestmentFromFriendsAmount Investors  
0                         0.0       258  
1                         0.0         1  
2                         0.0        41  
3                         0.0       158  
4                         0.0        20  

[5 rows x 81 columns]

## Assess

In [3]:
df.describe()

ListingNumber           Term    BorrowerAPR   BorrowerRate  \
count   1.139370e+05  113937.000000  113912.000000  113937.000000   
mean    6.278857e+05      40.830248       0.218828       0.192764   
std     3.280762e+05      10.436212       0.080364       0.074818   
min     4.000000e+00      12.000000       0.006530       0.000000   
25%     4.009190e+05      36.000000       0.156290       0.134000   
50%     6.005540e+05      36.000000       0.209760       0.184000   
75%     8.926340e+05      36.000000       0.283810       0.250000   
max     1.255725e+06      60.000000       0.512290       0.497500   

         LenderYield  EstimatedEffectiveYield  EstimatedLoss  EstimatedReturn  \
count  113937.000000             84853.000000   84853.000000     84853.000000   
mean        0.182701                 0.168661       0.080306         0.096068   
std         0.074516                 0.068467       0.046764         0.030403   
min        -0.010000                -0.182700       0.004900        -0.182700   
25%         0.124200                 0.115670       0.042400         0.074080   
50%         0.173000                 0.161500       0.072400         0.091700   
75%         0.240000                 0.224300       0.112000         0.116600   
max         0.492500                 0.319900       0.366000         0.283700   

       ProsperRating (numeric)  ProsperScore      ...        LP_ServiceFees  \
count             84853.000000  84853.000000      ...         113937.000000   
mean                  4.072243      5.950067      ...            -54.725641   
std                   1.673227      2.376501      ...             60.675425   
min                   1.000000      1.000000      ...           -664.870000   
25%                   3.000000      4.000000      ...            -73.180000   
50%                   4.000000      6.000000      ...            -34.440000   
75%                   5.000000      8.000000      ...            -13.920000   
max                   7.000000     11.000000      ...             32.060000   

       LP_CollectionFees  LP_GrossPrincipalLoss  LP_NetPrincipalLoss  \
count      113937.000000          113937.000000        113937.000000   
mean          -14.242698             700.446342           681.420499   
std           109.232758            2388.513831          2357.167068   
min         -9274.750000             -94.200000          -954.550000   
25%             0.000000               0.000000             0.000000   
50%             0.000000               0.000000             0.000000   
75%             0.000000               0.000000             0.000000   
max             0.000000           25000.000000         25000.000000   

       LP_NonPrincipalRecoverypayments  PercentFunded  Recommendations  \
count                    113937.000000  113937.000000    113937.000000   
mean                         25.142686       0.998584         0.048027   
std                         275.657937       0.017919         0.332353   
min                           0.000000       0.700000         0.000000   
25%                           0.000000       1.000000         0.000000   
50%                           0.000000       1.000000         0.000000   
75%                           0.000000       1.000000         0.000000   
max                       21117.900000       1.012500        39.000000   

       InvestmentFromFriendsCount  InvestmentFromFriendsAmount      Investors  
count               113937.000000                113937.000000  113937.000000  
mean                     0.023460                    16.550751      80.475228  
std                      0.232412                   294.545422     103.239020  
min                      0.000000                     0.000000       1.000000  
25%                      0.000000                     0.000000       2.000000  
50%                      0.000000                     0.000000      44.000000  
75%                      0.000000                     0.000000     

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 113937 entries, 0 to 113936
Data columns (total 81 columns):
ListingKey                             113937 non-null object
ListingNumber                          113937 non-null int64
ListingCreationDate                    113937 non-null object
CreditGrade                            28953 non-null object
Term                                   113937 non-null int64
LoanStatus                             113937 non-null object
ClosedDate                             55089 non-null object
BorrowerAPR                            113912 non-null float64
BorrowerRate                           113937 non-null float64
LenderYield                            113937 non-null float64
EstimatedEffectiveYield                84853 non-null float64
EstimatedLoss                          84853 non-null float64
EstimatedReturn                        84853 non-null float64
ProsperRating (numeric)                84853 non-null float64
ProsperRating (Alpha) 

In [5]:
df[df.BorrowerAPR.isnull()].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 25 entries, 2543 to 113921
Data columns (total 81 columns):
ListingKey                             25 non-null object
ListingNumber                          25 non-null int64
ListingCreationDate                    25 non-null object
CreditGrade                            25 non-null object
Term                                   25 non-null int64
LoanStatus                             25 non-null object
ClosedDate                             25 non-null object
BorrowerAPR                            0 non-null float64
BorrowerRate                           25 non-null float64
LenderYield                            25 non-null float64
EstimatedEffectiveYield                0 non-null float64
EstimatedLoss                          0 non-null float64
EstimatedReturn                        0 non-null float64
ProsperRating (numeric)                0 non-null float64
ProsperRating (Alpha)                  0 non-null object
ProsperScore         

In [6]:
df[df.DelinquenciesLast7Years.isnull()].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 990 entries, 206 to 113921
Data columns (total 81 columns):
ListingKey                             990 non-null object
ListingNumber                          990 non-null int64
ListingCreationDate                    990 non-null object
CreditGrade                            990 non-null object
Term                                   990 non-null int64
LoanStatus                             990 non-null object
ClosedDate                             990 non-null object
BorrowerAPR                            965 non-null float64
BorrowerRate                           990 non-null float64
LenderYield                            990 non-null float64
EstimatedEffectiveYield                0 non-null float64
EstimatedLoss                          0 non-null float64
EstimatedReturn                        0 non-null float64
ProsperRating (numeric)                0 non-null float64
ProsperRating (Alpha)                  0 non-null object
ProsperSco

In [7]:
pd.set_option('display.max_columns',81)
df.describe()

ListingNumber           Term    BorrowerAPR   BorrowerRate  \
count   1.139370e+05  113937.000000  113912.000000  113937.000000   
mean    6.278857e+05      40.830248       0.218828       0.192764   
std     3.280762e+05      10.436212       0.080364       0.074818   
min     4.000000e+00      12.000000       0.006530       0.000000   
25%     4.009190e+05      36.000000       0.156290       0.134000   
50%     6.005540e+05      36.000000       0.209760       0.184000   
75%     8.926340e+05      36.000000       0.283810       0.250000   
max     1.255725e+06      60.000000       0.512290       0.497500   

         LenderYield  EstimatedEffectiveYield  EstimatedLoss  EstimatedReturn  \
count  113937.000000             84853.000000   84853.000000     84853.000000   
mean        0.182701                 0.168661       0.080306         0.096068   
std         0.074516                 0.068467       0.046764         0.030403   
min        -0.010000                -0.182700       0.004900        -0.182700   
25%         0.124200                 0.115670       0.042400         0.074080   
50%         0.173000                 0.161500       0.072400         0.091700   
75%         0.240000                 0.224300       0.112000         0.116600   
max         0.492500                 0.319900       0.366000         0.283700   

       ProsperRating (numeric)  ProsperScore  ListingCategory (numeric)  \
count             84853.000000  84853.000000              113937.000000   
mean                  4.072243      5.950067                   2.774209   
std                   1.673227      2.376501                   3.996797   
min                   1.000000      1.000000                   0.000000   
25%                   3.000000      4.000000                   1.000000   
50%                   4.000000      6.000000                   1.000000   
75%                   5.000000      8.000000                   3.000000   
max                   7.000000     11.000000                  20.000000   

       EmploymentStatusDuration  CreditScoreRangeLower  CreditScoreRangeUpper  \
count             106312.000000          113346.000000          113346.000000   
mean                  96.071582             685.567731             704.567731   
std                   94.480605              66.458275              66.458275   
min                    0.000000               0.000000              19.000000   
25%                   26.000000             660.000000             679.000000   
50%                   67.000000             680.000000             699.000000   
75%                  137.000000             720.000000             739.000000   
max                  755.000000             880.000000             899.000000   

       CurrentCreditLines  OpenCreditLines  TotalCreditLinespast7years  \
count       106333.000000    106333.000000               113240.000000   
mean            10.317192         9.260164                   26.754539   
std              5.457866         5.022644                   13.637871   
min              0.000000         0.000000                    2.000000   
25%              7.000000         6.000000                   17.000000   
50%             10.000000         9.000000                   25.000000   
75%             13.000000        12.000000                   35.000000   
max             59.000000        54.000000                  136.000000   

       OpenRevolvingAccounts  OpenRevolvingMonthlyPayment  \
count           113937.00000                113937.000000   
mean                 6.96979                   398.292161   
std                  4.63097                   447.159711   
min                  0.00000                     0.000000   
25%                  4.00000                   114.000000   
50%                  6.00000                   271.000000   
75%                  9.00000                   525.000000   
max                 51.00000                 14985.000000   

       InquiriesLast6Months  Tot

In [8]:
df.Occupation.unique()

array(['Other', 'Professional', 'Skilled Labor', 'Executive',
       'Sales - Retail', 'Laborer', 'Food Service', 'Fireman',
       'Waiter/Waitress', 'Construction', 'Computer Programmer',
       'Sales - Commission', 'Retail Management', 'Engineer - Mechanical',
       'Military Enlisted', 'Clerical', nan, 'Teacher', 'Clergy',
       'Accountant/CPA', 'Attorney', 'Nurse (RN)', 'Analyst',
       "Nurse's Aide", 'Investor', 'Realtor', 'Flight Attendant',
       'Nurse (LPN)', 'Military Officer', 'Food Service Management',
       'Truck Driver', 'Administrative Assistant',
       'Police Officer/Correction Officer', 'Social Worker',
       'Tradesman - Mechanic', 'Medical Technician', 'Professor',
       'Postal Service', 'Civil Service', 'Pharmacist',
       'Tradesman - Electrician', 'Scientist', 'Dentist',
       'Engineer - Electrical', 'Architect', 'Landscaping',
       'Tradesman - Carpenter', 'Bus Driver', 'Tradesman - Plumber',
       'Engineer - Chemical', 'Doctor', 'Chemist',


In [9]:
df.EmploymentStatus.unique()

array(['Self-employed', 'Employed', 'Not available', 'Full-time', 'Other',
       nan, 'Not employed', 'Part-time', 'Retired'], dtype=object)

In [10]:
df.GroupKey.unique()

array([nan, '783C3371218786870A73D20', 'EF543365873906104DC9A04',
       '6C2E3384236387493493CB6', '67033364510034901E33BC7',
       'AD9A3400384972497EDAD38', '52EA3425051368132B80C96',
       'B5F93364617561572C32B6D', 'EE3833815612811323B64D2',
       '5EA73366320555283BA0246', '3D4D3366260257624AB272D',
       'A0CB3365552651755022A2E', 'B7BA33816687475120A8407',
       '156C3413203271953DEACE5', '4707336438924330015F446',
       '53B43365132049036C341A5', '6A3B336601725506917317E',
       'CC8D33653247904019A9059', '53B233889590051997E1E14',
       '5AFC33643897547591BD1C8', 'F555336594467179889CEFB',
       'B0473364376920128370B13', '4AC5336492870256755D273',
       'E3F433654060844916DB34F', 'DF643377029116179E7EDA3',
       'BAE93365625072437A1CBD9', 'FEF83377364176536637E50',
       '6CE133644751358832D593B', 'A8113366311579615424760',
       '85943386927308216031C0F', '25CB3382578620681398C96',
       'C80D337515684920818AB06', '214933649218625265C08F1',
       '29D03365204

In [11]:
df.FirstRecordedCreditLine.head()

0    2001-10-11 00:00:00
1    1996-03-18 00:00:00
2    2002-07-27 00:00:00
3    1983-02-28 00:00:00
4    2004-02-20 00:00:00
Name: FirstRecordedCreditLine, dtype: object

In [12]:
df[df.ListingKey.duplicated(keep=False)].sort_values(by='ListingKey').head(8)

ListingKey  ListingNumber            ListingCreationDate  \
32680  00223594917038064A7C947         998257  2013-11-15 16:58:37.167000000   
32681  00223594917038064A7C947         998257  2013-11-15 16:58:37.167000000   
32964  00473590513960687DD308F         941296  2013-10-07 15:47:36.023000000   
17274  00473590513960687DD308F         941296  2013-10-07 15:47:36.023000000   
7478   0098360461900952056DB93        1190614  2014-03-02 14:21:39.583000000   
33220  0098360461900952056DB93        1190614  2014-03-02 14:21:39.583000000   
27677  01163604029146842E28D9C        1233732  2014-02-25 14:33:39.830000000   
788    01163604029146842E28D9C        1233732  2014-02-25 14:33:39.830000000   

      CreditGrade  Term LoanStatus ClosedDate  BorrowerAPR  BorrowerRate  \
32680         NaN    36    Current        NaN      0.23540        0.1980   
32681         NaN    36    Current        NaN      0.23540        0.1980   
32964         NaN    60    Current        NaN      0.22242        0.1980   
17274         NaN    60    Current        NaN      0.22242        0.1980   
7478          NaN    36    Current        NaN      0.29660        0.2574   
33220         NaN    36    Current        NaN      0.29660        0.2574   
27677         NaN    36    Current        NaN      0.31975        0.2800   
788           NaN    36    Current        NaN      0.31975        0.2800   

       LenderYield  EstimatedEffectiveYield  EstimatedLoss  EstimatedReturn  \
32680       0.1880                  0.17416         0.0849          0.08926   
32681       0.1880                  0.17416         0.0849          0.08926   
32964       0.1880                  0.17416         0.0849          0.08926   
17274       0.1880                  0.17416         0.0849          0.08926   
7478        0.2474                  0.22530         0.1325          0.09280   
33220       0.2474                  0.22530         0.1325          0.09280   
27677       0.2700                  0.24550         0.1475          0.09800   
788         0.2700                  0.24550         0.1475          0.09800   

       ProsperRating (numeric) ProsperRating (Alpha)  ProsperScore  \
32680                      4.0                     C           5.0   
32681                      4.0                     C           6.0   
32964                      4.0                     C           4.0   
17274                      4.0                     C           6.0   
7478                       2.0                     E           2.0   
33220                      2.0                     E           4.0   
27677                      2.0                     E           2.0   
788                        2.0                     E           3.0   

       ListingCategory (numeric) BorrowerState    Occupation EmploymentStatus  \
32680                          1            NY         Other         Employed   
32681                          1            NY         Other         Employed   
32964                          1            TX    Nurse (RN)         Employed   
17274                          1            TX    Nurse (RN)         Employed   
7478                           1            VA  Professional         Employed   
33220                          1            VA  Professional         Employed   
27677                          1            NE         Other         Employed   
788                            1            NE         Other         Employed   

       EmploymentStatusDuration  IsBorrowerHomeowner  CurrentlyInGroup  \
32680                      69.0                False             False   
32681                      69.0                False             False   
32964                     218.0                 True             False   
17274                     218.0                 True             False   
7478                        4.0                False             False   
33220                       4.0                False             False   
27677                  

In [13]:
df.IncomeRange.unique()

array(['$25,000-49,999', '$50,000-74,999', 'Not displayed', '$100,000+',
       '$75,000-99,999', '$1-24,999', 'Not employed', '$0'], dtype=object)

In [14]:
df.query('ProsperScore > 10')

ListingKey  ListingNumber            ListingCreationDate  \
9       0F043596202561788EA13D5        1023355  2013-12-02 10:43:39.117000000   
69      0F37360159310530849DA23        1162592  2014-01-25 12:07:54.537000000   
158     108E36027739898797E47EC        1202850  2014-02-12 16:31:25.340000000   
224     2F293599751524472C82DEA        1130508  2014-01-10 07:24:44.853000000   
236     11D335932344051959C94C6         973657  2013-11-02 01:29:09.810000000   
277     2EFC36028887817232507BC        1220331  2014-02-18 12:59:08.680000000   
476     00EC3604528464581EB9190        1244304  2014-03-06 08:54:42.840000000   
491     32123598588424160C218A4        1116282  2014-01-06 07:18:54.173000000   
532     3BC83601275934190566C97        1150992  2014-01-20 10:51:29.120000000   
554     091B3596837609725239239        1054630  2013-12-17 22:15:59.470000000   
633     07FB35970598962377DCEA7        1079885  2013-12-15 12:25:42.193000000   
690     3CE83593217028064B104AB         991235  2013-11-03 08:48:26.140000000   
765     0A883595605120998B2479B        1022836  2013-12-02 07:36:43.197000000   
771     3D6435937004844127FA1D1         993658  2013-11-13 19:46:34.570000000   
854     46DB3596355549260D55E24        1082097  2013-12-12 11:08:56.630000000   
890     322936002397657303F6DAD        1137096  2014-01-14 09:02:45.567000000   
901     0A8F3595201741766C09C95        1026799  2013-12-03 00:36:12.727000000   
1017    354D3603799650128E20F1E        1236380  2014-03-06 03:45:19.167000000   
1085    3FBA35963912353000DA0E6        1079667  2013-12-12 03:37:30.163000000   
1153    42E13597205647526F425AD        1086246  2013-12-13 01:35:00.870000000   
1460    0B183592852839153D6FDB8         998145  2013-10-23 07:34:42.750000000   
1842    367E36032997310233C2E1E        1192663  2014-03-03 10:18:08.083000000   
1892    8AD035972820668966D958E        1054235  2013-12-05 12:28:21.723000000   
1962    3A1735899908154891E6951         911082  2013-09-19 11:47:58.917000000   
1972    54253591475644005A1085D         968697  2013-10-09 15:26:16.583000000   
2048    0BBA35970395345349717E3        1052408  2013-12-05 09:18:30.973000000   
2097    5E1A35989739796889A57BD        1069756  2013-12-31 10:34:15.973000000   
2181    03E53599590858524748DAB        1107074  2013-12-30 16:58:06.427000000   
2288    2AA136039431268891F2466        1166956  2014-02-19 14:11:22.497000000   
2487    04C63595369815186D0A35C        1026919  2013-12-05 04:50:07.360000000   
...                         ...            ...                            ...   
111771  2C493599702647240352564        1125892  2014-01-21 14:48:16.600000000   
111890  2AEA358216796526862A4CF         827053  2013-07-02 14:09:36.273000000   
111938  BFD03593083358631FC9A44        1005026  2013-11-09 05:52:10.357000000   
112024  30FE3595795974847E869A2        1014469  2013-11-22 18:22:02.563000000   
112098  B7B135986438897720D55AC        1086346  2014-01-06 06:25:23.533000000   
112124  D53E3590146755144C437AA         912223  2013-09-19 12:44:10.020000000   
112316  C4B03596594624959D4B5BB        1056467  2013-12-05 22:05:04.813000000   
112549  DE573594409510502FBB362        1018934  2013-11-14 19:24:22.833000000   
112646  CD7335938639285316D87BA         979285  2013-11-04 14:12:13.310000000   
112653  FE1F3602658538034A1A258        1146712  2014-02-07 21:40:58.917000000   
112684  DC9C359295492959916CD02        1027866  2013-11-08 05:04:32.807000000   
112774  E4D53598980813958C72AF5        1091854  2014-01-07 14:18:59.403000000   
112775  DF6F3602534484466AC80E8        1169617  2014-02-19 20:32:03.660000000   
112808  D02A359764366391596EB30        1040752  2013-12-09 13:56:51.750000000   
112857  E52E3594003437610E20626         971240  2013-10-28 02:32:23.193000000   
112909  D2A23596462014942814780        1032395  2013-11-25 08:06:35.157000000   
112921  E40D3593544866386EB5142         970853  2013-10-24 19:39:25.217000000   
112990  D55635921380677717B3959         950327  2013

In [26]:
df.CreditGrade.unique()

array(['C', nan, 'HR', 'AA', 'D', 'B', 'E', 'A', 'NC'], dtype=object)

### Issues
#### Quality:
1. 25 rows missing BorrowerAPR  and other fields
2. 990 rows missing deliquency related information
3. ListingCreationDate,ClosedDate, LoanOriginationDate, FirstRecordedCreditLine DateCreditPulled are not datetimes
4. CreditGrade, LoanStatus, ProsperRating (numeric), ProsperRating (alpha), Occupation, EmploymentStatus, ListingCategory (numeric), IncomeRange,  are not categorical
5. Lots of duplicate rows with only prosper rating changing


## Cleaning

### Datapoints missing many fields
#### Define:
Remove rows missing BorrowerAPR or DelinquenciesLast7Years 
#### Code:

In [15]:
df.dropna(subset = ['DelinquenciesLast7Years','BorrowerAPR'], inplace=True)

#### Test:

In [16]:
df[['DelinquenciesLast7Years','BorrowerAPR']].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 112947 entries, 0 to 113936
Data columns (total 2 columns):
DelinquenciesLast7Years    112947 non-null float64
BorrowerAPR                112947 non-null float64
dtypes: float64(2)
memory usage: 2.6 MB


### Duplicate Rows
#### Define:
Remove duplicate rows

#### Code:

In [18]:
df.drop_duplicates(subset='ListingKey', inplace=True)

#### Test:

In [19]:
df[df.ListingKey.duplicated()]

Empty DataFrame
Columns: [ListingKey, ListingNumber, ListingCreationDate, CreditGrade, Term, LoanStatus, ClosedDate, BorrowerAPR, BorrowerRate, LenderYield, EstimatedEffectiveYield, EstimatedLoss, EstimatedReturn, ProsperRating (numeric), ProsperRating (Alpha), ProsperScore, ListingCategory (numeric), BorrowerState, Occupation, EmploymentStatus, EmploymentStatusDuration, IsBorrowerHomeowner, CurrentlyInGroup, GroupKey, DateCreditPulled, CreditScoreRangeLower, CreditScoreRangeUpper, FirstRecordedCreditLine, CurrentCreditLines, OpenCreditLines, TotalCreditLinespast7years, OpenRevolvingAccounts, OpenRevolvingMonthlyPayment, InquiriesLast6Months, TotalInquiries, CurrentDelinquencies, AmountDelinquent, DelinquenciesLast7Years, PublicRecordsLast10Years, PublicRecordsLast12Months, RevolvingCreditBalance, BankcardUtilization, AvailableBankcardCredit, TotalTrades, TradesNeverDelinquent (percentage), TradesOpenedLast6Months, DebtToIncomeRatio, IncomeRange, IncomeVerifiable, StatedMonthlyIncome, LoanKey, TotalProsperLoans, TotalProsperPaymentsBilled, OnTimeProsperPayments, ProsperPaymentsLessThanOneMonthLate, ProsperPaymentsOneMonthPlusLate, ProsperPrincipalBorrowed, ProsperPrincipalOutstanding, ScorexChangeAtTimeOfListing, LoanCurrentDaysDelinquent, LoanFirstDefaultedCycleNumber, LoanMonthsSinceOrigination, LoanNumber, LoanOriginalAmount, LoanOriginationDate, LoanOriginationQuarter, MemberKey, MonthlyLoanPayment, LP_CustomerPayments, LP_CustomerPrincipalPayments, LP_InterestandFees, LP_ServiceFees, LP_CollectionFees, LP_GrossPrincipalLoss, LP_NetPrincipalLoss, LP_NonPrincipalRecoverypayments, PercentFunded, Recommendations, InvestmentFromFriendsCount, InvestmentFromFriendsAmount, Investors]
Index: []

### Datetimes
#### Define:
Correct data type of dates
#### Code:

In [20]:
dates = ['ListingCreationDate','ClosedDate', 'LoanOriginationDate', 'FirstRecordedCreditLine', 'DateCreditPulled']
df[dates] = df[dates].apply(pd.to_datetime)

#### Test:

In [21]:
df[dates].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 112076 entries, 0 to 113936
Data columns (total 5 columns):
ListingCreationDate        112076 non-null datetime64[ns]
ClosedDate                 54086 non-null datetime64[ns]
LoanOriginationDate        112076 non-null datetime64[ns]
FirstRecordedCreditLine    112076 non-null datetime64[ns]
DateCreditPulled           112076 non-null datetime64[ns]
dtypes: datetime64[ns](5)
memory usage: 5.1 MB


In [24]:
df[dates].head()

ListingCreationDate ClosedDate LoanOriginationDate  \
0 2007-08-26 19:09:29.263 2009-08-14          2007-09-12   
1 2014-02-27 08:28:07.900        NaT          2014-03-03   
2 2007-01-05 15:00:47.090 2009-12-17          2007-01-17   
3 2012-10-22 11:02:35.010        NaT          2012-11-01   
4 2013-09-14 18:38:39.097        NaT          2013-09-20   

  FirstRecordedCreditLine        DateCreditPulled  
0              2001-10-11 2007-08-26 18:41:46.780  
1              1996-03-18 2014-02-27 08:28:14.000  
2              2002-07-27 2007-01-02 14:09:10.060  
3              1983-02-28 2012-10-22 11:02:32.000  
4              2004-02-20 2013-09-14 18:38:44.000

### Categories 
#### Define: 
Change categorical data to the correct data type

#### Code:

CreditGrade, LoanStatus, ProsperRating (numeric), ProsperRating (alpha), Occupation, EmploymentStatus, ListingCategory (numeric), IncomeRange

In [30]:
#Sequentials
cols = ['CreditGrade', 'LoanStatus', 'ProsperRating (numeric)', 'ProsperRating (Alpha)', 'Occupation', 
       'EmploymentStatus', 'ListingCategory (numeric)', 'IncomeRange']
df[cols] = df[cols].astype('category')

#### Test:

In [32]:
df[cols].dtypes

CreditGrade                  category
LoanStatus                   category
ProsperRating (numeric)      category
ProsperRating (Alpha)        category
Occupation                   category
EmploymentStatus             category
ListingCategory (numeric)    category
IncomeRange                  category
dtype: object

### Explore: